tests/conftest.py

```diff
@@ -0,0 +1,231 @@
+       parametrized test functions. If string is specified, the objects will
+       be shred in the test functions that have same 'shared_result' value.
+       If you don't specify specific_srcdir, this option override
+       specific_srcdir param by 'shared_result' value. Default is None.
+    """
+    env = request.node.get_marker('testenv')
+    kwargs = env.kwargs if env else {}
+    result = {
+       'build': False,  # pre build in fixture
+        'specific_srcdir': None,
+        'shared_result': None,
+    }
+    result.update(kwargs)
+
+    if (result['shared_result'] and
+            not isinstance(result['shared_result'], string_types)):
+        r = result['shared_result'] = request.node.originalname or request.node.name
+
+    if result['shared_result'] and not result['specific_srcdir']:
+        result['specific_srcdir'] = result['shared_result']
+
+    if (result['specific_srcdir'] and
+        not isinstance(result['specific_srcdir'], string_types)):
+        result['specific_srcdir'] = request.node.originalname or request.node.name
+
+    return result
+
+
+@pytest.fixture(scope='function')
+def app(test_params, app_params, make_app, shared_result):
+    """
+    provides sphinx.application.Sphinx object
+    """
+    args, kwargs = app_params
+    if test_params['specific_srcdir'] and 'srcdir' not in kwargs:
+        kwargs['srcdir'] = test_params['specific_srcdir']
+
+    if test_params['shared_result']:
+        restore = shared_result.restore(test_params['shared_result'])
+        kwargs.update(restore)
+
+    app_ = make_app(*args, **kwargs)
+
+    if test_params['build']:
+        # if listdir is not empty, we can use built cache
+        if not app_.outdir.listdir():
+            app_.build()
+    yield app_
+
+    if test_params['shared_result']:
+        shared_result.store(test_params['shared_result'], app_)
+
+
+@pytest.fixture(scope='function')
+def status(app):
+    """
+    compat for testing with previous @with_app decorator
+    """
+    return app._status
+
+
+@pytest.fixture(scope='function')
+def warning(app):
+    """
+    compat for testing with previous @with_app decorator
+    """
+    return app._warning
+
+
+@pytest.fixture()
+def make_app():
+    """
+    provides make_app function to initialize SphinxTestApp instance.
+    if you want to initialize 'app' in your test function. please use this
+    instead of using SphinxTestApp class directory.
+    """
+    apps = []
+    syspath = sys.path[:]
+
+    def make(*args, **kwargs):
+        status, warning = StringIO(), StringIO()
+        kwargs.setdefault('status', status)
+        kwargs.setdefault('warning', warning)
+        app_ = SphinxTestApp(*args, **kwargs)
+        apps.append(app_)
+        return app_
+    yield make
+
+    sys.path[:] = syspath
+    for app_ in apps:
+        app_.cleanup()
+
+
+class SharedResult(object):
+    cache = {}
+
+    def store(self, key, app_):
```

Comment:
The name of `app` is used as a fixture at module level.
So I want to avoid overriding the name.